src/main/java/apoc/util/Util.java

```diff
@@ -686,4 +691,99 @@ public static boolean classExists(String className) {
             return null;
         }
     }
+
+    public static Stream<MapResult> runWithRetry( GraphDatabaseService db, Log log, String statement, long retries, List<String> additionalRetryCodes,
+            Map<String,Object> params )
+    {
+        String transactionFailureMessage = "";
+        Long timesTried = 0L;
+        do
+        {
+
+            FutureTask<Result> resultFuture = new FutureTask<>( () -> {
+                try ( Transaction tx = db.beginTx() )
+                {
+                    Result result = ((Function<Map<String,Object>,Result>) ( Map<String,Object> p ) -> db.execute( withParamMapping( statement, p.keySet() ),
+                            p )).apply( params );
+                    tx.success();
+                    return result;
+                }
+                catch ( Exception e )
+                {
+                    throw e;
+                }
+            } );
+            Thread threadForTransaction = new Thread( resultFuture );
+
+            try
+            {
+                threadForTransaction.start();
+                threadForTransaction.join();
+
+                // Query has not finished yet, wait.
+                while ( !resultFuture.isDone() )
+                {
+                    LockSupport.parkNanos( 1000 );
+                }
+
+                return resultFuture.get().stream().map( MapResult::new );
+            }
+            catch ( Exception e )
+            {
+                Throwable cause = e.getCause();
+
+                if ( cause instanceof QueryExecutionException )
+                {
+
+                    String statusCode = ((QueryExecutionException) cause).getStatusCode();
+                    if ( !isRetriableException( statusCode, additionalRetryCodes ) )
+                    {
+                        transactionFailureMessage = "Failed after " + timesTried + " out of " + retries + " retries. QueryExecutionException : " +
+                                ((QueryExecutionException) cause).getStatusCode();
+                        break;
+                    }
+                }
+
+                log.warn( "Retrying operation " + timesTried + " of " + retries );
+                Util.sleep( 100 );
+                timesTried++;
+            }
+            transactionFailureMessage = "Failed after " + timesTried + " of " + retries + " retries.";
+        }
+        while ( timesTried <= retries );
+
+        throw new TransactionFailureException( transactionFailureMessage );
+    }
+
+    private interface RetriableExceptions extends Status
+    {
+        List<Status> QueryExecutionExceptionStatusCodeList = Arrays.asList(
```

Comment:
why is this in an interface?
did you check if status has equals perhaps a set with contains is good enough?

Reply:
we should utilize Code.classification for transient errrors
https://github.com/neo4j/neo4j/blob/3.5/community/common/src/main/java/org/neo4j/kernel/api/exceptions/Status.java#L80-L91

Reply:
@jexp It is an interface since the inner enums of Status are `protected`. And I did check the Status interface, it does not have the exceptions grouped in a nice way for what we would want to retry on. We want to have control over which exceptions can cause a retry.

What were you thinking/meaning by utilizing `code.classification` for transient errors?